analyzers/src/SonarAnalyzer.Common/Rules/UseIFormatProviderForParsingDateAndTimeBase.cs

```diff
@@ -0,0 +1,79 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Globalization;
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseIFormatProviderForParsingDateAndTimeBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S6580";
+
+    private static readonly string[] parseMethodNames = new[]
+    {
+        nameof(DateTime.Parse),
+        nameof(DateTime.ParseExact),
+        nameof(DateTime.TryParse),
+        nameof(DateTime.TryParseExact)
+    };
+    private static readonly string[] nonDeterministicFormatProviders = new[]
```

Comment:
I am not 100% sure we have to cover this case. It might be some very weird case where the string is generated on the same machine with the same Culture and consumed as is too. In that case, there is no problem. And I would assume once the dev provides an IFormatProvider it is a conscious decision.